network/p2pNetwork.go

```diff
@@ -0,0 +1,456 @@
+		n.wsPeersConnectivityCheckTicker.Stop()
+		n.wsPeersConnectivityCheckTicker = nil
+	}
+	n.ctxCancel()
+	n.service.Close()
+	n.wg.Wait()
+}
+
+func (n *P2PNetwork) meshThread() {
+	defer n.wg.Done()
+	timer := time.NewTicker(meshThreadInterval)
+	defer timer.Stop()
+	for {
+		select {
+		case <-timer.C:
+			n.service.DialPeers(n.config.GossipFanout)
+		case <-n.ctx.Done():
+			return
+		}
+	}
+}
+
+// GetGenesisID implements GossipNode
+func (n *P2PNetwork) GetGenesisID() string {
+	return n.genesisID
+}
+
+// Address returns a string and whether that is a 'final' address or guessed.
+func (n *P2PNetwork) Address() (string, bool) {
+	addrInfo := peer.AddrInfo{
+		ID:    n.service.Host().ID(),
+		Addrs: n.service.Host().Addrs(),
+	}
+	addrs, err := peer.AddrInfoToP2pAddrs(&addrInfo)
+	if err != nil {
+		n.log.Warnf("Failed to generate valid multiaddr: %v", err)
+		return "", false
+	}
+	if len(addrs) == 0 {
+		return "", false
+	}
+	if len(addrs) > 1 {
+		n.log.Infof("Multiple addresses found, using first one from %v", addrs)
+	}
+
+	return addrs[0].String(), true
+}
+
+// Broadcast sends a message.
+func (n *P2PNetwork) Broadcast(ctx context.Context, tag protocol.Tag, data []byte, wait bool, except Peer) error {
+	// For tags using pubsub topics, publish to GossipSub
+	if topic, ok := n.topicTags[tag]; ok {
+		return n.service.Publish(ctx, topic, data)
+	}
+	// Otherwise broadcast over websocket protocol stream
+	return n.broadcaster.BroadcastArray(ctx, []protocol.Tag{tag}, [][]byte{data}, wait, except)
+}
+
+// Relay message
+func (n *P2PNetwork) Relay(ctx context.Context, tag protocol.Tag, data []byte, wait bool, except Peer) error {
+	return n.Broadcast(ctx, tag, data, wait, except)
+}
+
+// Disconnect from a peer, probably due to protocol errors.
+func (n *P2PNetwork) Disconnect(badnode Peer) {
+	switch node := badnode.(type) {
+	case peer.ID:
+		err := n.service.Host().Network().ClosePeer(node)
+		if err != nil {
+			n.log.Warnf("Error disconnecting from peer %s: %v", node, err)
+		}
+	default:
+		n.log.Warnf("Unknown peer type %T", badnode)
+	}
+}
+
+func (n *P2PNetwork) disconnectThread(badnode Peer, reason disconnectReason) {
+	defer n.wg.Done()
+	n.Disconnect(badnode) // ignores reason
+}
+
+// DisconnectPeers is used by testing
+func (n *P2PNetwork) DisconnectPeers() {
+	for _, conn := range n.service.Host().Network().Conns() {
+		conn.Close()
+	}
+}
+
+// RegisterHTTPHandler path accepts gorilla/mux path annotations
+func (n *P2PNetwork) RegisterHTTPHandler(path string, handler http.Handler) {
+}
+
+// RequestConnectOutgoing asks the system to actually connect to peers.
+// `replace` optionally drops existing connections before making new ones.
+// `quit` chan allows cancellation.
+func (n *P2PNetwork) RequestConnectOutgoing(replace bool, quit <-chan struct{}) {
+}
+
+// GetPeers returns a list of Peers we could potentially send a direct message to.
+func (n *P2PNetwork) GetPeers(options ...PeerOption) []Peer { return nil }
```

Comment:
will this be implemented eventually to support catchpoint catchup?

Reply:
Potentially yeah, right now it's just left trivial to implement the GossipNode interface.